File: bench/json/json.go
// json.go — general JSON parser (encoding/json into map[string]interface{}) for benchmark.
package main

import (
	"encoding/json"
	"fmt"
	"time"
)

const src = `{"items":[{"id":1,"kind":2,"value":10},{"id":2,"kind":3,"value":20},{"id":3,"kind":5,"value":30}],"meta":{"scale":7,"bias":11}}`

const (
	nIters  = 512
	nRuns   = 21
	nWarmup = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

// getInt extracts an int32 from a float64 value in the generic map
// (encoding/json unmarshals all numbers as float64 in interface{} context).
func getInt(v interface{}) int32 {
	if f, ok := v.(float64); ok {
		return int32(int64(f))
	}
	return 0
}

func walk() uint32 {
	h := uint32(0x811c9dc5)
	for i := 0; i < nIters; i++ {
		// Parse into a fully generic value — no schema knowledge.
		var root map[string]interface{}
		_ = json.Unmarshal([]byte(src), &root)

		// Walk by string-key access on the generic map.
		items := root["items"].([]interface{})
		meta  := root["meta"].(map[string]interface{})
		scale := getInt(meta["scale"])
		s     := getInt(meta["bias"])

		for _, elem := range items {
			it := elem.(map[string]interface{})
			id    := getInt(it["id"])
			kind  := getInt(it["kind"])
			value := getInt(it["value"])
			s += id*scale + kind + value
		}
		h = mix(h, uint32(s))
	}
	return h
}

func main() {
	var cs uint32
	for i := 0; i < nWarmup; i++ {
		cs = walk()
	}

	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		cs = walk()
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), cs, nIters, 4, nRuns)
}

File: bench/json/json.js
import { medianUs, mix, printResult } from '../_lib/benchlib.js'

const SRC = '{"items":[{"id":1,"kind":2,"value":10},{"id":2,"kind":3,"value":20},{"id":3,"kind":5,"value":30}],"meta":{"scale":7,"bias":11}}'
const N_ITERS = 512
const N_RUNS = 21
const N_WARMUP = 5

const walk = () => {
  let h = 0x811c9dc5 | 0
  for (let i = 0; i < N_ITERS; i++) {
    const o = JSON.parse(SRC)
    const items = o.items
    let s = o.meta.bias
    for (let j = 0; j < items.length; j++) {
      const it = items[j]
      s += it.id * o.meta.scale + it.kind + it.value
    }
    h = mix(h, s)
  }
  return h >>> 0
}

export let main = () => {
  let cs = 0
  for (let i = 0; i < N_WARMUP; i++) cs = walk()

  const samples = new Float64Array(N_RUNS)
  for (let i = 0; i < N_RUNS; i++) {
    const t0 = performance.now()
    cs = walk()
    samples[i] = performance.now() - t0
  }
  printResult(medianUs(samples), cs, N_ITERS, 4, N_RUNS)
}
